pkg/usagetracker/tracker_test.go

```diff
@@ -3,15 +3,355 @@
+		require.NoError(t, a1.reconcilePartitions(context.Background()))
+
+		// zone-a-0 should have 8 partitions, zone-a-1 should have 8, this is deterministic at this point.
+		a0.partitionsMtx.RLock()
+		require.Len(t, a0.partitions, 8)
+		a0.partitionsMtx.RUnlock()
+		a1.partitionsMtx.RLock()
+		require.Len(t, a1.partitions, 8)
+		a1.partitionsMtx.RUnlock()
+
+		// Check eventually that owners are updated in the ring.
+		require.EventuallyWithT(t, func(t *assert.CollectT) {
+			for partitionID := int32(0); partitionID < testPartitionsCount; partitionID++ {
+				require.Len(t, partitionRing.PartitionRing().MultiPartitionOwnerIDs(partitionID, nil), 2, "Partition %d should have 2 owners.", partitionID)
+			}
+		}, 5*lostPartitionsShutdownGracePeriod, 100*time.Millisecond)
+	})
+
+	t.Run("respects MaxPartitionsToCreatePerReconcile", func(t *testing.T) {
+		t.Parallel()
+
+		ikv, pkv, cluster := prepareKVStoreAndKafkaMocks(t)
+		const (
+			expectedZoneAReconcilesToCreateAllPartitions = 4
+			expectedZoneBReconcilesToCreateAllPartitions = 2
+		)
+		trackers := map[string]*UsageTracker{
+			"a0": newTestUsageTracker(t, 0, "zone-a", ikv, pkv, cluster, func(cfg *Config) {
+				cfg.MaxPartitionsToCreatePerReconcile = testPartitionsCount / expectedZoneAReconcilesToCreateAllPartitions
+			}),
+			"b0": newTestUsageTracker(t, 0, "zone-b", ikv, pkv, cluster, func(cfg *Config) {
+				cfg.MaxPartitionsToCreatePerReconcile = testPartitionsCount / expectedZoneBReconcilesToCreateAllPartitions
+			}),
+		}
+		waitUntilAllTrackersSeeAllInstancesInTheirZones(t, trackers)
+
+		// Reconcile once:
+		for key, ut := range trackers {
+			require.NoError(t, ut.reconcilePartitions(context.Background()), "Tracker %q could not reconcile", key)
+			ut.partitionsMtx.RLock()
+			require.Equal(t, ut.cfg.MaxPartitionsToCreatePerReconcile, len(ut.partitions))
+			ut.partitionsMtx.RUnlock()
+		}
+
+		// Should not be ready yet:
+		requireAllTrackersNotReady(t, trackers)
+
+		// Reconcile rest expected times.
+		for reconcile := 1; reconcile < expectedZoneAReconcilesToCreateAllPartitions; reconcile++ {
+			require.NoError(t, trackers["a0"].reconcilePartitions(context.Background()), "Tracker a0 could not reconcile")
+		}
+		for reconcile := 1; reconcile < expectedZoneBReconcilesToCreateAllPartitions; reconcile++ {
+			require.NoError(t, trackers["b0"].reconcilePartitions(context.Background()), "Tracker b0 could not reconcile")
+		}
+
+		// Should be ready now.
+		for key, ut := range trackers {
+			require.NoError(t, ut.CheckReady(context.Background()), "Tracker %q should be ready now", key)
+		}
+	})
+}
+
+func requireAllTrackersReady(t *testing.T, trackers map[string]*UsageTracker) {
+	t.Helper()
+	for key, ut := range trackers {
+		require.NoError(t, ut.CheckReady(context.Background()), "Tracker %q should be ready now", key)
+	}
+}
+
+func requireAllTrackersNotReady(t *testing.T, trackers map[string]*UsageTracker) {
```

Comment:
Should this and next functions be a `t.Helper()` too?

Reply:
Nah, that doesn't help. `t.Helper()` would hide this function from the erro trace, but it can fail in many places, and we really want to see where it fails.